services/src/kilda-pce/src/main/java/org/openkilda/pce/AvailableNetworkFactory.java

```diff
@@ -105,14 +108,36 @@ public AvailableNetwork getAvailableNetwork(
 
     private Collection<Isl> getAvailableIsls(BuildStrategy buildStrategy, Flow flow) {
         if (buildStrategy == BuildStrategy.COST) {
-            return flow.isIgnoreBandwidth() ? islRepository.findAllActive() :
+            Collection<Isl> isls = flow.isIgnoreBandwidth() ? islRepository.findAllActive() :
                     islRepository.findActiveWithAvailableBandwidth(flow.getBandwidth(), flow.getEncapsulationType());
+            validateIslsCost(isls);
+            return isls;
+
         } else if (buildStrategy == BuildStrategy.SYMMETRIC_COST) {
-            return flow.isIgnoreBandwidth() ? islRepository.findAllActive() :
+            Collection<Isl> isls = flow.isIgnoreBandwidth() ? islRepository.findAllActive() :
                     islRepository.findSymmetricActiveWithAvailableBandwidth(flow.getBandwidth(),
                             flow.getEncapsulationType());
+            validateIslsCost(isls);
+            return isls;
         } else {
-            throw new UnsupportedOperationException(String.format("Unsupported buildStrategy type %s", buildStrategy));
+            throw new UnsupportedOperationException(format("Unsupported buildStrategy type %s", buildStrategy));
+        }
+    }
+
+    private void validateIslsCost(Collection<Isl> isls) {
+        List<Isl> islsWithNegativeCost = isls.stream()
+                .filter(isl -> isl.getCost() < 0)
+                .collect(Collectors.toList());
+
+        if (!islsWithNegativeCost.isEmpty()) {
+            List<String> islMessages = new ArrayList<>();
+            for (Isl isl : islsWithNegativeCost) {
```

Comment:
Why don't you build the list with islMessages during the first iteration over isls? The current solution builds excessive intermediate list of isls with negative costs.